storage: use atomic replication changes in RelocateRange

tbg opened the pull request:
I wrote this up rather quickly, but wanted to get this out for review sooner
rather than later.

----

Touches #12768.

Release justification: the previous code would enter vulnerable
replication configurations when it wasn't necessary, thus undermining
what we wanted to achieve in #12768.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/41084)
<!-- Reviewable:end -->

tbg commented on pkg/storage/replica_command.go:
```diff
@@ -2009,15 +2010,118 @@ func (s *Store) AdminRelocateRange(
+			transferLease(*leaseTarget)
+		}
+
+		ctx = context.WithValue(ctx, "testing", "testing")
```
TODO: Need to fix this hack; without it AdminChangeReplicas will run the changes one by one.

tbg replied:
Ok, I did something.

danhhz reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> more straightforward that I was thinking it would be

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-:-LpdReN00xNNy17LYXBm:b-fnnbjr)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @danhhz and @tbg)

___
*[pkg/storage/client_relocate_range_test.go, line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpdMJjp9A2T62g9Uah0:-LpdMJjp9A2T62g9Uah1:bcwzuad) ([raw file](https://github.com/cockroachdb/cockroach/blob/58f384173b7e57b29bc0fe96d6255576f5ec588a/pkg/storage/client_relocate_range_test.go#L36)):*
> ```Go
> 	require.NoError(t, tc.Servers[0].DB().AdminRelocateRange(ctx, startKey.AsRawKey(), targets))
> 	var desc roachpb.RangeDescriptor
> 	require.NoError(t, tc.Servers[0].DB().GetProto(
> ```

why not `tc.Servers[0].LookupRange`?
___
*[pkg/storage/client_relocate_range_test.go, line 168 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpdP2AR58QrqDUTCjm1:-LpdP2ASCHLEK3s-8BtN:b-jmxdwf) ([raw file](https://github.com/cockroachdb/cockroach/blob/58f384173b7e57b29bc0fe96d6255576f5ec588a/pkg/storage/client_relocate_range_test.go#L168)):*
> ```Go
> 	// A grab bag.
> 	{
> 		// s3 -(add)-> s3 s2 -(swap)-> s4 s2 -(add)-> s4 s2 s1
> ```

nit: the last part of this comment doesn't match
___
*[pkg/storage/client_relocate_range_test.go, line 172 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpdP9DwE1OwN_5fUP5-:-LpdP9DwE1OwN_5fUP50:b-jmxdwf) ([raw file](https://github.com/cockroachdb/cockroach/blob/58f384173b7e57b29bc0fe96d6255576f5ec588a/pkg/storage/client_relocate_range_test.go#L172)):*
> ```Go
> 			relocateAndCheck(t, tc, k, tc.Targets(1, 3, 0))
> 		})
> 		// s2 s4 s1 -(add)-> s2 s4 s1 s6
> ```

nit: the last part of this comment doesn't match
___
*[pkg/storage/replica_command.go, line 2020 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpdJjH83iN85qFS1CKQ:-LpdJjH83iN85qFS1CKR:b-xr5q3c) ([raw file](https://github.com/cockroachdb/cockroach/blob/58f384173b7e57b29bc0fe96d6255576f5ec588a/pkg/storage/replica_command.go#L2020)):*
> ```Go
> 		}
> 		for _, substr := range whitelist {
> 			if strings.Contains(err.Error(), substr) {
> ```

nit: probably better to be defensive and call `err.Error()` once
___
*[pkg/storage/replica_command.go, line 2064 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpdK8SG3J_V_dGBOcex:-LpdK8SG3J_V_dGBOcey:b-x1jd8g) ([raw file](https://github.com/cockroachdb/cockroach/blob/58f384173b7e57b29bc0fe96d6255576f5ec588a/pkg/storage/replica_command.go#L2064)):*
> ```Go
> 		}
> 
> 		ops, leaseTarget, err := s.relocateOne(ctx, &rangeDesc, targets)
> ```

check this err
___
*[pkg/storage/replica_command.go, line 2077 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpdKN7G6me1uNhLiNG1:-LpdKN7G6me1uNhLiNG2:bqxb4z1) ([raw file](https://github.com/cockroachdb/cockroach/blob/58f384173b7e57b29bc0fe96d6255576f5ec588a/pkg/storage/replica_command.go#L2077)):*
> ```Go
> 
> 		ctx = context.WithValue(ctx, "testing", "testing")
> 		newDesc, err := s.DB().AdminChangeReplicas(
> ```

nit: seems like this should fit on one line
___
*[pkg/storage/replica_command.go, line 2085 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpdKZMNBJmx2pv7syi8:-LpdKZMODtw-wuKddBRl:bxife9e) ([raw file](https://github.com/cockroachdb/cockroach/blob/58f384173b7e57b29bc0fe96d6255576f5ec588a/pkg/storage/replica_command.go#L2085)):*
> ```Go
> 			}
> 			if every.ShouldLog() {
> 				log.Warning(ctx, returnErr)
> ```

Warning seems a bit high for this given the "descriptor changed" stuff
___
*[pkg/storage/replica_command.go, line 2087 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpdKg9Y9EhR6SJIrbl5:-LpdKg9Y9EhR6SJIrbl6:b3dwsed) ([raw file](https://github.com/cockroachdb/cockroach/blob/58f384173b7e57b29bc0fe96d6255576f5ec588a/pkg/storage/replica_command.go#L2087)):*
> ```Go
> 				log.Warning(ctx, returnErr)
> 			}
> 			re.Next()
> ```

nit: move this into the `for {`?
___
*[pkg/storage/replica_command.go, line 2105 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpdKxpx5XctH6CoCeq8:-LpdKxpy2Ca45Kse_jNY:bmbja29) ([raw file](https://github.com/cockroachdb/cockroach/blob/58f384173b7e57b29bc0fe96d6255576f5ec588a/pkg/storage/replica_command.go#L2105)):*
> ```Go
> 	rangeReplicas := desc.Replicas().All()
> 	if len(rangeReplicas) != len(desc.Replicas().Voters()) {
> 		return nil, nil, crdberrors.AssertionFailedf(
> ```

looks like you lost the comment pointing out why this is safe to assert
___
*[pkg/storage/replica_command.go, line 2247 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpdNzwQEzeVMZ03xw6A:-LpdNzwQEzeVMZ03xw6B:b-rzcw7q) ([raw file](https://github.com/cockroachdb/cockroach/blob/58f384173b7e57b29bc0fe96d6255576f5ec588a/pkg/storage/replica_command.go#L2247)):*
> ```Go
> 		liReq.Key = desc.StartKey.AsRawKey()
> 		b.AddRawRequest(liReq)
> 		if err := s.DB().Run(ctx, &b); err != nil {
> ```

we don't already have a helper for this?
___
*[pkg/storage/replica_command.go, line 2012 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpdRVOg1Wymfm-GSJZ8:-LpdRVOg1Wymfm-GSJZ9:bp6tlre) ([raw file](https://github.com/cockroachdb/cockroach/blob/801cc5cdc3d3e3b1f1ccf80137fdf105fc9830e4/pkg/storage/replica_command.go#L2012)):*
> ```Go
> 		//
> 		// TODO(tbg): remove in 20.1.
> 		ctx = context.WithValue(ctx, "testing", "testing")
> ```

we should update this value name now that's it's not just testing
___
*[pkg/storage/replica_command.go, line 2091 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpdQ5D-6WSwHHbcTa1j:-LpdQ5D07sTjSNftYfmo:bar7ufg) ([raw file](https://github.com/cockroachdb/cockroach/blob/801cc5cdc3d3e3b1f1ccf80137fdf105fc9830e4/pkg/storage/replica_command.go#L2091)):*
> ```Go
> 		}
> 		// When a swap is in order but we don't have atomic replication changes,
> 		// run the ops one by one.
> ```

do we need this? `ChangeReplicas` seems to unroll them


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
TFTR! This is basically ready to be merged I think, but I made enough changes to wait for another pass. I've also noticed that the test I added sometimes takes ~1.5s but other times 11s.. looking into why, it seems that there are unexplained pauses sometimes, such as this one:

```
I190926 09:07:22.951158 4200 storage/replica_command.go:2075  [n4,s4,r25/9:/{Table/Max-Max}] TBG ====== LOOP START
I190926 09:07:22.951197 4200 storage/replica_command.go:2077  [n4,s4,r25/9:/{Table/Max-Max}]        TBG ====== ATTEMPT START
I190926 09:07:22.952095 4303 storage/replica_command.go:1583  [n3,s3,r25/12:/{Table/Max-Max}] change replicas (add [] remove [(n2,s2):8]): existing descriptor r25:/{Table/Max-Max} [(n5,s5):13, (n2,s2):8, (n3,s3):12, (n6,s6):11, next=14, gen=34, sticky=9223372036.854775807,2147483647]
I190926 09:07:25.843486 4303 storage/replica_raft.go:291  [n3,s3,r25/12:/{Table/Max-Max}] proposing REMOVE_REPLICA[(n2,s2):8]: after=[(n5,s5):13 (n6,s6):11 (n3,s3):12] next=14
I190926 09:07:25.844452 1120 storage/store.go:2660  [n2,s2,r25/8:/{Table/Max-Max}] removing replica r25/8
```

I'm going to spend some time later today figuring out what those are about, there's probably an opportunity to make replication more snappy across the board.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-:-LpgpJiVBoMgAGb8nTSS:bggjdd6)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @danhhz)

___
*[pkg/storage/client_relocate_range_test.go, line 168 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpdP2AR58QrqDUTCjm1:-LpgerpA43nNNPKoCp1O:bhyo0y2) ([raw file](https://github.com/cockroachdb/cockroach/blob/58f384173b7e57b29bc0fe96d6255576f5ec588a/pkg/storage/client_relocate_range_test.go#L168)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: the last part of this comment doesn't match
</blockquote></details>

Added the reordered forms (I intentionally didn't reorder so that each step would be easier to follow along with).
___
*[pkg/storage/replica_command.go, line 2085 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpdKZMNBJmx2pv7syi8:-LpggBgDEWwhNgfSMmVX:b-lyj8m0) ([raw file](https://github.com/cockroachdb/cockroach/blob/58f384173b7e57b29bc0fe96d6255576f5ec588a/pkg/storage/replica_command.go#L2085)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Warning seems a bit high for this given the "descriptor changed" stuff
</blockquote></details>

That's the old code, but I agree. Downgraded.
___
*[pkg/storage/replica_command.go, line 2087 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpdKg9Y9EhR6SJIrbl5:-LpggMXo6ZdGBqCqgElt:bcsy3qm) ([raw file](https://github.com/cockroachdb/cockroach/blob/58f384173b7e57b29bc0fe96d6255576f5ec588a/pkg/storage/replica_command.go#L2087)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: move this into the `for {`?
</blockquote></details>

I rearranged this a bit -- the for loop was both "do a thing at a time until no more things need doing" and "retry the current thing with a backoff". Pulled those apart so we now have one more layer of for loops and more idiomatic-looking backoff code.
___
*[pkg/storage/replica_command.go, line 2247 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpdNzwQEzeVMZ03xw6A:-LpgiIRwEytCnTe6r4CX:b-f662nf) ([raw file](https://github.com/cockroachdb/cockroach/blob/58f384173b7e57b29bc0fe96d6255576f5ec588a/pkg/storage/replica_command.go#L2247)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

we don't already have a helper for this?
</blockquote></details>

Not that I'm aware, I've seen helpers only in the testcluster machinery.
___
*[pkg/storage/replica_command.go, line 2091 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpdQ5D-6WSwHHbcTa1j:-LpgfJ-XDKcw16KWoJRj:b-59h3ga) ([raw file](https://github.com/cockroachdb/cockroach/blob/801cc5cdc3d3e3b1f1ccf80137fdf105fc9830e4/pkg/storage/replica_command.go#L2091)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

do we need this? `ChangeReplicas` seems to unroll them
</blockquote></details>

No, added a comment

		// When a swap is in order but we're not sure that all nodes are running
		// 19.2+ (in which the AdminChangeReplicas RPC was extended to support
		// mixing additions and removals), don't send such requests but unroll
		// the ops here, running them one by one. 19.2 nodes that don't have the
		// cluster version enabled would unroll them internally, but 19.1 nodes
		// don't even understand the RPC and would interpret all ops as
		// additions (which would fail since we can't add an existing node).


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
Ok, I figured it out. TL;DR is that we sometimes end up removing the raft leader and then we have to wait out the 3s election timeout. Filed https://github.com/cockroachdb/cockroach/pull/41122 to fix this.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-:-LphLyt75K2V8bTGlV4w:b-fyinhw)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @danhhz)



<!-- Sent from Reviewable.io -->

danhhz reviewed (commented):
still <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-:-LpiBnPEDIoIL5RkvMqA:b-5fsfav)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @danhhz and @tbg)

___
*[pkg/storage/replica_command.go, line 2091 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpdQ5D-6WSwHHbcTa1j:-Lpi9qRZ74iOG_AetDfv:b-1likf) ([raw file](https://github.com/cockroachdb/cockroach/blob/801cc5cdc3d3e3b1f1ccf80137fdf105fc9830e4/pkg/storage/replica_command.go#L2091)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

No, added a comment

		// When a swap is in order but we're not sure that all nodes are running
		// 19.2+ (in which the AdminChangeReplicas RPC was extended to support
		// mixing additions and removals), don't send such requests but unroll
		// the ops here, running them one by one. 19.2 nodes that don't have the
		// cluster version enabled would unroll them internally, but 19.1 nodes
		// don't even understand the RPC and would interpret all ops as
		// additions (which would fail since we can't add an existing node).
</blockquote></details>

Aha, yes, good to have this comment
___
*[pkg/storage/replica_command.go, line 2075 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpiBa3c0Z6ToCjoolB_:-LpiBa3c0Z6ToCjoolBa:b-2yddho) ([raw file](https://github.com/cockroachdb/cockroach/blob/b317bd4afd35a9906e830645f1981408db1d6652/pkg/storage/replica_command.go#L2075)):*
> ```Go
> 	every := log.Every(time.Minute)
> 	for {
> 		for re := retry.StartWithCtx(ctx, retry.Options{MaxBackoff: 5 * time.Second}); ; re.Next() {
> ```

I suspect some of the control flow gets easier to follow (breaks and success bool etc) if this inner loop moves into a method


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 1 of 3 files at r1, 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41084)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @danhhz and @tbg)

___
*[pkg/internal/client/db.go, line 564 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpiGdEfDVEV4S_GpzaJ:-LpiGdEfDVEV4S_GpzaK:b-sav4vs) ([raw file](https://github.com/cockroachdb/cockroach/blob/b317bd4afd35a9906e830645f1981408db1d6652/pkg/internal/client/db.go#L564)):*
> ```Go
> 
> // ChangeReplicasCanMixAddAndRemoveContext convinces
> // (*client.DB).AdminChangeReplicas of that the caller is aware that 19.1 nodes
> ```

s/of //
___
*[pkg/storage/client_relocate_range_test.go, line 108 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpiKfAA2RTrX3Wb1nWU:-LpiKfAA2RTrX3Wb1nWV:bv7gmxn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b317bd4afd35a9906e830645f1981408db1d6652/pkg/storage/client_relocate_range_test.go#L108)):*
> ```Go
> 				intercepted = append(intercepted, intercept{
> 					ops:         ops,
> 					leaseTarget: leaseTarget,
> ```

We don't seem to ever check this. Should we be? I don't see us checking the leaseholder anywhere else, but the comments imply that we know who it will become.
___
*[pkg/storage/client_relocate_range_test.go, line 138 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpiKMoAB1D0RZiRa5Ci:-LpiKMoAB1D0RZiRa5Cj:b-5aqsa8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b317bd4afd35a9906e830645f1981408db1d6652/pkg/storage/client_relocate_range_test.go#L138)):*
> ```
> swap out replicas one by one
> ```

Isn't this doing the opposite? Or do you mean that it's performing three atomic replica changes, one by one?
___
*[pkg/storage/replica_command.go, line 2087 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpiJ1Xn3IAc_9F7VIAY:-LpiJ1Xn3IAc_9F7VIAZ:bf0kau3) ([raw file](https://github.com/cockroachdb/cockroach/blob/b317bd4afd35a9906e830645f1981408db1d6652/pkg/storage/replica_command.go#L2087)):*
> ```Go
> 				// Done.
> 				// Step 3: Transfer the lease to the first listed target, as the API specifies.
> 				transferLease(targets[0])
> ```

Should we just make `relocateOne` return this as the `leaseTarget` in this case?
___
*[pkg/storage/replica_command.go, line 2103 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpiIKG7DX3FjiV_Ylqd:-LpiIKG7DX3FjiV_Ylqe:bovx748) ([raw file](https://github.com/cockroachdb/cockroach/blob/b317bd4afd35a9906e830645f1981408db1d6652/pkg/storage/replica_command.go#L2103)):*
> ```Go
> 
> 			opss := [][]roachpb.ReplicationChange{ops}
> 			if !useAtomic && len(ops) == 2 {
> ```

Why only if `len(ops) == 2`? If I'm reading the comment above, we'd want to do this whenever `!useAtomic`.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
Got errors under stressrace related to verifying the lease placement which I fixed; seems stable on my gceworker. In 19.2 we should look into making sure the lease transfers become mandatory, right now relocate will just fail weirdly if any of them don't go through (but oddly they seem to have a 100% success rate).

Thanks for the reviews!

bors r=danhhz

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-:-LpmCLyN9NrUK6D3NFIK:bml2m3h)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @danhhz and @nvanbenschoten)

___
*[pkg/storage/client_relocate_range_test.go, line 108 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpiKfAA2RTrX3Wb1nWU:-Lplqi6nAgefM1k62aPu:b-7aitd6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b317bd4afd35a9906e830645f1981408db1d6652/pkg/storage/client_relocate_range_test.go#L108)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We don't seem to ever check this. Should we be? I don't see us checking the leaseholder anywhere else, but the comments imply that we know who it will become.
</blockquote></details>

The test verifies that the lease ends up in the right place (see `requireLeaseAt`), it's hard to predict the exact lease transfers in between since the allocator ultimately gets to decide what order things happen in, and the allocator tends to be... not deterministic.
___
*[pkg/storage/client_relocate_range_test.go, line 138 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpiKMoAB1D0RZiRa5Ci:-LplrhGbCZqScv1lnPFC:bks4tz2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b317bd4afd35a9906e830645f1981408db1d6652/pkg/storage/client_relocate_range_test.go#L138)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> ```
> swap out replicas one by one
> ```

Isn't this doing the opposite? Or do you mean that it's performing three atomic replica changes, one by one?
</blockquote></details>

Yep. Clarified the comment.
___
*[pkg/storage/replica_command.go, line 2075 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpiBa3c0Z6ToCjoolB_:-LplsEII-y5Jbtn_albJ:b-hxbtoj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b317bd4afd35a9906e830645f1981408db1d6652/pkg/storage/replica_command.go#L2075)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

I suspect some of the control flow gets easier to follow (breaks and success bool etc) if this inner loop moves into a method
</blockquote></details>

Maybe, but the `return err` paths get more awkward. I'm going to leave as is for now since the whole `opss` thing goes away in 19.2 and with it the weird control flow.
___
*[pkg/storage/replica_command.go, line 2103 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/41084#-LpiIKG7DX3FjiV_Ylqd:-LpluZP7EuZa7AoQN58v:b-egq3ii) ([raw file](https://github.com/cockroachdb/cockroach/blob/b317bd4afd35a9906e830645f1981408db1d6652/pkg/storage/replica_command.go#L2103)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why only if `len(ops) == 2`? If I'm reading the comment above, we'd want to do this whenever `!useAtomic`.
</blockquote></details>

`len(ops)==2` is either `1` or `2` and if it's `1` then we shouldn't make `opss` have an extra nil slice at the end. I added an assertion just above to make sure this is clear to readers.


<!-- Sent from Reviewable.io -->

tbg committed:
storage: use atomic replication changes in RelocateRange

Touches #12768.

Release justification: the previous code would enter vulnerable
replication configurations when it wasn't necessary, thus undermining
what we wanted to achieve in #12768.

PS: the newly added test passes when manually overriding useAtomic with
`false`, and I verified that it indeed uses no atomic replication
changes.

Release note: None

tbg commented:
Linters were unhappy

bors r=danhhz
